File: 2416. Sum of Prefix Scores of Strings.py
# Definition of the TrieNode class
class TrieNode:
    def __init__(self):
        self.count = 0                     # Initialize the count of prefix occurrences to 0
        self.children = [None] * 26        # Initialize the children array with 26 None (for 'a' to 'z')

# Definition of the Solution class
class Solution:
    def sumPrefixScores(self, words: list[str]) -> list[int]:
        root = TrieNode()  # Create the root of the Trie

        # 1. Insert all words into the Trie and count the occurrence of each prefix
        for w in words:
            node = root                     # Start at the root for each word
            for ch in w:
                index = ord(ch) - ord('a')  # Convert character to its corresponding index (0 to 25)
                if not node.children[index]: # If the corresponding TrieNode doesn't exist, create a new one
                    node.children[index] = TrieNode()
                node = node.children[index] # Move to the child node
                node.count += 1             # Increment the count for this prefix

        # 2. Calculate the sum of prefix scores for each word
        ans = []  # List to store the prefix scores for each word

        for w in words:
            sum_score = 0                   # Variable to store the sum of counts for the current word
            node = root                     # Start at the root again
            for ch in w:
                index = ord(ch) - ord('a')  # Convert character to index
                node = node.children[index] # Move to the corresponding child node
                sum_score += node.count     # Add the count of this prefix to the sum
            ans.append(sum_score)           # Store the result in the answer list

        return ans                          # Return the final result

File: 2416. Sum of Prefix Scores of Strings.cpp
// Definition of a TrieNode
class TrieNode
{
public:
    int count;              // This will store the count of words sharing the current prefix
    TrieNode *children[26]; // Array to hold pointers to the next TrieNode for each letter (26 lowercase letters)

    // Constructor
    TrieNode() : count(0), children{NULL} {} // Initialize count to 0, and all children pointers to NULL (nullptr)
};

class Solution
{
public:
    vector<int> sumPrefixScores(vector<string> &words)
    {
        // Create the root of the Trie
        auto *root = new TrieNode();

        // 1. Insert each word into the Trie and count the occurrence of each prefix
        for (auto w : words)
        {                      // Iterate through each word in the list of words
            auto *node = root; // Start from the root node for each word

            // Insert each character of the word into the Trie
            for (auto ch : w)
            {                         // For each character in the word:
                int assci = ch - 'a'; // Convert the character to its index ('a' = 0, 'b' = 1, ..., 'z' = 25)

                // If the corresponding TrieNode doesn't exist, create a new node
                if (!node->children[assci])
                    node->children[assci] = new TrieNode();

                // Move to the next TrieNode (child)
                node = node->children[assci];

                // Increment the count for this prefix (each prefix is counted as it appears)
                node->count += 1;
            }
        }

        // 2. Calculate the sum of prefix scores for each word
        vector<int> ans; // Vector to store the result (prefix sums for each word)

        // Traverse through the words again to compute their prefix scores
        for (auto w : words)
        {                      // For each word in the list:
            auto sum = 0;      // Variable to store the sum of prefix counts for this word
            auto *node = root; // Start from the root node

            // Traverse through the word in the Trie to compute the sum
            for (auto ch : w)
            {                                 // For each character in the word:
                int assci = ch - 'a';         // Convert the character to its index
                node = node->children[assci]; // Move to the corresponding child node in the Trie

                // Add the current prefix's count to the sum
                sum += node->count;
            }

            // Store the sum of prefix scores for this word in the result vector
            ans.push_back(sum);
        }

        // Return the final result vector containing prefix sums for all words
        return ans;
    }
};
